### ISSUE TITLE
Attempt to convert from mariadb to postgresql fails with unicode errors

### ISSUE THREAD

**Issue by rkarlsba** [znuny/Znuny#164]
- OS: Debian Buster
- Znuny version 6.0.37

### Expected behaviour

works

### Actual behaviour

[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Code point 0x1DDECE is not Unicode, may not be portable in print at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.
[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Unicode surrogate U+DA81 is illegal in UTF-8 at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.
[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Code point 0x13C3DA is not Unicode, may not be portable in print at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.
[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Unicode surrogate U+DFB3 is illegal in UTF-8 at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.
[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Code point 0x15A475 is not Unicode, may not be portable in print at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.
[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Unicode surrogate U+DF7E is illegal in UTF-8 at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.
[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Code point 0x1D5208 is not Unicode, may not be portable in print at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.
[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Code point 0x17C8F7 is not Unicode, may not be portable in print at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.
[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Unicode surrogate U+DCD7 is illegal in UTF-8 at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.
[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Unicode surrogate U+DFA7 is illegal in UTF-8 at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.
[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Unicode surrogate U+DFD8 is illegal in UTF-8 at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.
[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Code point 0x12CDBF is not Unicode, may not be portable in print at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.
[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Code point 0x13B4BC is not Unicode, may not be portable in print at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.
[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Code point 0x1BAA8E is not Unicode, may not be portable in print at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.
[Tue Nov  2 17:25:36 2021] otrs.Console.pl: Code point 0x1DDECE is not Unicode, may not be portable in print at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 322.

### How to reproduce

./bin/otrs.Console.pl Maint::Database::Clone --dry-run

### Additional information

I've moved the 300GB worth of articles from file storage to db storage on mariadb, giving me a large performance improvement, but I want to move it to our postgres cluster to see if that can help out, both in terms of performance and also, giving me a chance for redundancy. This seems to fail as seen above.

**Comment by rkaldung** [comment]
Hi @rkarlsba,

What's your MariaDB character and collation set for the database and what do you use for PostgreSQL?

**Comment by rkarlsba** [comment]
Excuse my ignorance, but for clarity, how do I check that?

**Comment by rkaldung** [comment]
MariaDB: `SELECT default_character_set_name,default_collation_name FROM information_schema.SCHEMATA  WHERE schema_name = "znuny"` Replace znuny with the database name
PostgreSQL: I use the cli client and command `\l`:
```

**Comment by rkarlsba** [comment]
MariaDB [(none)]> SELECT default_character_set_name,default_collation_name FROM information_schema.SCHEMATA WHERE schema_name = "otrs02";

+----------------------------+------------------------+
| default_character_set_name | default_collation_name |
+----------------------------+------------------------+
| utf8                       | utf8_general_ci        |
+----------------------------+------------------------+

 znuny02             | znuny02          | UTF8     | en_US.UTF-8 | en_US.UTF-8 |

**Comment by rkarlsba** [comment]
Also, this is a perl error (or warning?), not database. Any idea how to fix this?

**Comment by rkarlsba** [comment]
I tried to start a migration anyway to see if it worked. It didn't.

[Wed Nov  3 15:57:27 2021] otrs.Console.pl: DBD::Pg::db do failed: ERROR:  invalid byte sequence for encoding "UTF8": 0xed 0xa0 0xbd at /opt/otrs/Kernel/System/DB.pm line 471.
ERROR: OTRS-otrs.Console.pl-Maint::Database::Clone-64 Perl: 5.32.1 OS: linux Time: Wed Nov 3 15:57:27 2021

 Message: ERROR:  invalid byte sequence for encoding "UTF8": 0xed 0xa0 0xbd, SQL: 'INSERT INTO article_data_mime (id, a_from, a_reply_to, a_to, a_cc, a_subject, a_message_id, a_in_reply_to, a_references, a_content_type, a_body, incoming_time, content_path, create_time, create_by, change_time, change_by, a_message_id_md5, article_id, a_bcc) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)'

 Traceback (49063):
   Module: Kernel::System::CloneDB::Driver::Base::DataTransfer Line: 380
   Module: Kernel::System::CloneDB::Backend::DataTransfer Line: 228
   Module: Kernel::System::Console::Command::Maint::Database::Clone::Run Line: 108
   Module: (eval) Line: 461
   Module: Kernel::System::Console::BaseCommand::Execute Line: 455
   Module: Kernel::System::Console::InterfaceConsole::Run Line: 81
   Module: ./bin/otrs.Console.pl Line: 37

Could not insert data: Table: article_data_mime - id:844184.
Error: Died at /opt/otrs/Kernel/System/CloneDB/Driver/Base.pm line 386.
otrs@znuny02:~$

**Comment by rkaldung** [comment]
@rkarlsba This seems to be a left over from the storage migration which was not migrated because the record is broken. You should check the current database and try to fix the record. This is nothing what CloneDB could handle.

**Comment by rkarlsba** [comment]
What sort of tool would you recommend to use to repair the db?

**Comment by hanneshal** [comment]
I dont think that any tool can repair this. The inserted data are corrupt / not valid. 

This is an older entry?

**Comment by rkarlsba** [comment]
I don't know. I haven't examined it closely enough. I don't even know where to start looking. It seems mariadb just ignores these sort of errors and has probably filled up the db with tons of crap. So - it works on mariadb and although I really want to move it to postgres, well, I guess it probably won't be worth the time.

**Comment by rkaldung** [comment]
I guess you ran the storage switch with the parameter `--tolerant` which exactly ignores the issue you have. There are many possibilities what could be the cause of this encoding issue. I would check the table article_data_mime for the broken records (manually) to identify them and the check them in the UI. If you don't need them or it's spam delete the ticket and try a migration again.

**Comment by rkarlsba** [comment]
I did indeed run it with --tolerant - it didn't succeed without it. Any idea how to do this properly?

**Comment by rkarlsba** [comment]
How can I run it without --tolerant and somewhat delete or flag the invalid articles? Manually going through them will be quite impossible.

**Comment by hanneshal** [comment]
You could just redirect the output to an textfile and get all the article_id / attachment ids. 
With this list you can clean everything up / delete invalid attachments.

**Comment by rkarlsba** [comment]
> You could just redirect the output to an textfile and get all the article_id / attachment ids. With this list you can clean everything up / delete invalid attachments.

thanks, but can you please detail? Is this this you suggest?

$ ./bin/otrs.Console.pl Admin::Article::StorageSwitch --target ArticleStorageDB > /somewhere/big/somelog

or this?

$ ./bin/otrs.Console.pl Admin::Article::StorageSwitch --tolerant --target ArticleStorageDB > /somewhere/big/somelog

**Comment by rkaldung** [comment]
@rkarlsba Is the log file generated which is configure by the system configuration `CloneDB::LogFile` (default value CloneDB::LogFile)?

**Comment by rkarlsba** [comment]
Thanks for the reply. I'll look further into this over the holiday. All system freeze now.

**Comment by rkaldung** [comment]
@rkarlsba Any progress on this or can we close this issue?
